### PR TITLE
embeddable etcdmain

### PR DESCRIPTION
**Pull request by heyitsanthony**
Mostly a config / startEtcd refactor

### CODE REVIEW

File: etcdmain/etcd.go [hunk 1]
```diff
@@ -188,161 +189,215 @@ func startEtcdOrProxyV2() {
...
+		// TODO: https://github.com/golang/go/issues/9524
+		go func() {
+			e.errc <- sctx.serve(e.Server, ctlscfg, ch, e.errc)
+		}()
+	}
+	return nil
+}
+
+// Etcd contains a running etcd server and its listeners.
+type Etcd struct {
```
Comment (xiang90) [inline etcdmain/etcd.go]: is it possible to move this out of etcdmain? probably etcd/embed.Server?

**Comment by xiang90** [comment]
This is the direction I have in mind. Only comment is that we might move the etcd structure to somewhere else. etcdmain.Etcd seems strange to me. The original purpose of etcdmain is just to hold the main func. Now it grows...

File: etcdmain/config.go [hunk 2]
```diff
@@ -450,9 +513,9 @@ func initialClusterFromName(name string) string {
 	return fmt.Sprintf("%s=http://localhost:2380", n)
 }
 
-func (cfg config) isNewCluster() bool          { return cfg.clusterState.String() == clusterStateFlagNew }
+func (cfg Config) isNewCluster() bool          { return cfg.ClusterState == clusterStateFlagNew }
 func (cfg config) isProxy() bool               { return cfg.proxy.String() != proxyFlagOff }
```
Comment (gyuho) [inline etcdmain/config.go]: `cfg Config`?

Comment (heyitsanthony) [inline etcdmain/config.go]: still tearing this up; proxy config stuff will be private to `etcdmain`, regular etcd config stuff will be in `embed`

**Comment by purpleidea** [comment]
Thanks for looking into this. It's probably mentioning my earlier take on this in https://github.com/coreos/etcd/pull/5584

Let me know when this is passing and ready for review, I'll try refactoring my https://github.com/purpleidea/mgmt/ on top of it to get a feel for how it would work any any issues.

**Comment by xiang90** [comment]
@purpleidea This pr tries to achieve what you want in: https://github.com/coreos/etcd/pull/5584, what openshift (another embedded etcd user) needs, what kuberentes need, and what our friends from pingCAP needs. We figured it would be hard to ask you directly to coordinate with all the related people. So we decided to do it on our side. But we will try to make sure it achieves most of things you want. Thanks!

File: embed/config.go [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+	Debug        bool   `json:"debug"`
+	LogPkgLevels string `json:"log-package-levels"`
+	EnablePprof  bool
+
+	// ForceNewCluster starts a new cluster even if previously started; unsafe.
+	ForceNewCluster bool `json:"force-new-cluster"`
+}
+
+// config holds the config suitable for yaml parsing
+type config struct {
```
Comment (xiang90) [inline embed/config.go]: do we use this for any other purpose? if this is just for yaml, probably we should rename this to configYAML.

File: embed/config.go [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+	ClientTLSInfo transport.TLSInfo
+	ClientAutoTLS bool
+	PeerTLSInfo   transport.TLSInfo
+	PeerAutoTLS   bool
+
+	// debug
+
+	Debug        bool   `json:"debug"`
+	LogPkgLevels string `json:"log-package-levels"`
+	EnablePprof  bool
```
Comment (gyuho) [inline embed/config.go]: why do some fields have no json tag?

Comment (heyitsanthony) [inline embed/config.go]: either they didn't have json tags before or they need additional processing after being loaded from json (e.g., converting url strings into url.Urls)

Comment (gyuho) [inline embed/config.go]: got it. thanks

File: embed/config.go (hunk 3 above)
Comment (heyitsanthony) [inline embed/config.go]: sure

**Comment by xiang90** [comment]
@heyitsanthony Is the test failure related? seems like ctl_v3 failed. other than that, lgtm.

**Comment by heyitsanthony** [comment]
@xiang90 yeah, there's something going on with the peer connections so quorum e2e tests are failing; still debugging.

**Pull request merged**